pkg/registry/core/service/storage/rest.go

```diff
@@ -390,35 +433,32 @@ func (rs *REST) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 	}
 
 	// TODO: this should probably move to strategy.PrepareForCreate()
-	releaseServiceIP := false
+	var toReleaseClusterIPs map[api.IPFamily]string
+	var toAllocClusterIPs map[api.IPFamily]string
+
+	// failed allocation release
 	defer func() {
-		if releaseServiceIP {
-			if helper.IsServiceIPSet(service) {
-				allocator := rs.getAllocatorByClusterIP(service)
-				allocator.Release(net.ParseIP(service.Spec.ClusterIP))
-			}
+		// release the allocate .. allocate the released
+		released, errRelease := rs.releaseClusterIPs(toReleaseClusterIPs)
+		allocated, errAllocate := rs.allocClusterIPs(oldService, toAllocClusterIPs)
```

Comment:
that is why we have repair loop. The repair cleans up leaked or deallocated ips in case of failures. The code above has been trying to get it right but it can not cover _all_ failure scenarios.